mockagent_test.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2020 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package jaeger
+
+import (
+	"context"
+	"net"
+	"sync"
+	"testing"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/stretchr/testify/require"
+
+	"storj.io/monkit-jaeger/gen-go/agent"
+	"storj.io/monkit-jaeger/gen-go/jaeger"
+)
+
+// StartMockAgent starts a mock agent on a local udp port.
+func StartMockAgent(t *testing.T, f func(mock *MockAgent)) {
+	mock := NewMockAgent()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		err := mock.Serve()
+		require.NoError(t, err)
+		wg.Done()
+	}()
+
+	<-mock.started
```

Comment:
In cases like this it is better to not allow raw channel reads, as that can create unintentional deadlocks. I suggest using `sync2.Fence` for this. So something like:

```
type MockAgent struct {
    fence sync2.Fence
    ....
}

// WaitForStart waits until the mock agent has started serving
func (m *MockAgent) WaitForStart(ctx context.Context) {
    return m.fence.Wait(ctx)
}
```

Then in the serve function, you can call `m.fence.Release()`.

This will require you to add a ctx argument to `runWithAgent`, but that's a good idea either way.

Alternatively, if using a fence seems like overkill, when you simply want to signal something like this the best way is to just close a channel. Reads on closed channels immediately return.

So in that scenario you'd have a channel on `MockAgent` that signals start. I'd still have the `WaitForStart` method that simply reads on the channel, and then when `Serve` is ready it calls `close(waitCh)`.

Or a sync.WaitGroup. Lots of options. :) But let's have some type of API to expose this rather than reading the raw channel.

Reply:
I went with the simple one. Just wrapped the read into `WaitForStart` and using the `close` to signaling